drivers/dma/dma_silabs_siwx917.c

```diff
@@ -0,0 +1,602 @@
+		return -EINVAL;
+	}
+	channel_control.srcSize = siwx917_dma_data_width(config->source_data_size);
+	channel_control.dstSize = siwx917_dma_data_width(config->dest_data_size);
+	/* Validate the addr increment value */
+	if ((siwx917_dma_addr_adjustment(config->head_block->source_addr_adj) < 0) ||
+	    (siwx917_dma_addr_adjustment(config->head_block->dest_addr_adj) < 0)) {
+		return -EINVAL;
+	}
+	/* Update source and destination addr increment values */
+	if (siwx917_dma_addr_adjustment(config->head_block->source_addr_adj) == 0) {
+		channel_control.srcInc = channel_control.srcSize;
+	} else {
+		channel_control.srcInc = UDMA_SRC_INC_NONE;
+	}
+	if (siwx917_dma_addr_adjustment(config->head_block->dest_addr_adj) == 0) {
+		channel_control.dstInc = channel_control.dstSize;
+	} else {
+		channel_control.dstInc = UDMA_DST_INC_NONE;
+	}
+	/* Configure dma channel for transfer */
+	status = (sl_status_t)UDMAx_ChannelConfigure(
+		&UDMA_Resources, (uint8_t)rsi_channel,
+		(uint32_t)(config->head_block->source_address),
+		(uint32_t)(config->head_block->dest_address), config->head_block->block_size,
+		channel_control, &channel_config, NULL, channel_info, udma_handle);
+	return status;
+}
+
+/*Function to configure UDMA channel for transfer*/
+static int dma_siwx917_configure(const struct device *dev, uint32_t channel,
+				 struct dma_config *config)
+{
+
+	/* Fetch DMA config and data structures from device. */
+	const struct dma_siwx917_config *cfg = dev->config;
+	struct dma_siwx917_data *data = dev->data;
+
+	uint32_t dma_instance;
+	sl_status_t status;
+	/* channel info structure to temporarily hold the channel data. */
+	static UDMA_Channel_Info channel_info[32];
+	uint32_t rsi_channel = channel - 1;
+	RSI_UDMA_HANDLE_T udma_handle;
+
+	if (cfg->reg == UDMA0) {
+		dma_instance = UDMA0_INSTANCE; /* udma0 */
+		/* Expecting a fixed channel number between 1-32 */
+		if ((channel > DT_PROP(DT_NODELABEL(udma0), dma_channels)) || (channel == 0)) {
+			return -EINVAL;
+		}
+		udma_handle = siwx917_udma_handle[UDMA0_INSTANCE];
+	} else if (cfg->reg == UDMA1) {
+		dma_instance = ULP_UDMA_INSTANCE; /* ulp_udma */
+		/* Expecting a fixed channel number between 1-12 */
+		if ((channel > DT_PROP(DT_NODELABEL(udma1), dma_channels)) || (channel == 0)) {
+			return -EINVAL;
+		}
+		udma_handle = siwx917_udma_handle[ULP_UDMA_INSTANCE];
+	} else {
+		return -EINVAL;
+	}
+
+	/* Disable the channel before configuring */
+	if (RSI_UDMA_ChannelDisable(udma_handle, rsi_channel) != SL_STATUS_OK) {
+		return -ECANCELED;
+	}
+
+	/* Validate the priority */
+	if ((config->channel_priority != DMA_CH_PRIORITY_LOW) &&
+	    (config->channel_priority != DMA_CH_PRIORITY_HIGH)) {
+		LOG_ERR("Invalid DMA priority - must be low priority (0), high priority (1)");
+		return -EINVAL;
+	}
+
+	/* Allocate the channel for transfer */
+	status = sl_si91x_dma_allocate_channel(dma_instance, &channel, config->channel_priority);
```

Comment:
Do we need `sl_si91x_dma_allocate_channel()`? The Zephyr API do not support dynamic allocation of the DMA channels.

We are mixing 3 layers of API: `RSI_UDMA_*()`, `UDMAx_*()` and `sl_si91x_dma_*()`. If I understand right, `sl_si91x_dma_*()` relies on `dma_instance` which do not fit very well with Zephyr. Maybe you could get rid of it.

Reply:
Removed sl layer implementation